File: test/ts/ssh-test/keyImportExportTests.ts
//
//  Copyright (c) Microsoft Corporation. All rights reserved.
//

import * as assert from 'assert';
import { suite, test, params, pending, slow, timeout } from '@testdeck/mocha';
import {
	KeyFormat,
	KeyEncoding,
	keyFormatters,
	importKey,
	exportPublicKey,
	exportPrivateKey,
	exportPublicKeyBytes,
	importKeyBytes,
	exportPrivateKeyBytes,
	Pkcs8KeyFormatter,
} from '@microsoft/dev-tunnels-ssh-keys';
import { MockRandom } from './mockRandom';

@suite
@slow(200)
export class KeyImportExportTests {
	@test
	@params({ algorithm: 'rsa2048', keyFormat: KeyFormat.Ssh })
	@params({ algorithm: 'rsa2048', keyFormat: KeyFormat.Pkcs1 })
	@params({ algorithm: 'rsa2048', keyFormat: KeyFormat.Pkcs8 })
	@params({ algorithm: 'rsa2048', keyFormat: KeyFormat.Jwk })
	@params({ algorithm: 'rsa4096', keyFormat: KeyFormat.Ssh })
	@params({ algorithm: 'rsa4096', keyFormat: KeyFormat.Pkcs1 })
	@params({ algorithm: 'rsa4096', keyFormat: KeyFormat.Pkcs8 })
	@params({ algorithm: 'rsa4096', keyFormat: KeyFormat.Jwk })
	@params({ algorithm: 'ecdsa384', keyFormat: KeyFormat.Ssh })
	@params({ algorithm: 'ecdsa384', keyFormat: KeyFormat.Pkcs8 })
	@params({ algorithm: 'ecdsa384', keyFormat: KeyFormat.Jwk })
	@params({ algorithm: 'ecdsa521', keyFormat: KeyFormat.Ssh })
	@params({ algorithm: 'ecdsa521', keyFormat: KeyFormat.Pkcs8 })
	@params.naming((p) => `importPublicKey(${p.algorithm}, ${getFormatName(p.keyFormat)})`)
	public async importPublicKey({
		algorithm,
		keyFormat,
	}: {
		algorithm: string;
		keyFormat: KeyFormat;
	}) {
		const suffix = getFormatName(keyFormat);
		const keyPair = await importKey(readTestFile(`public-${algorithm}-${suffix}`));
		assert(keyPair);
		assert(keyPair.hasPublicKey);
		assert(!keyPair.hasPrivateKey);

		let expected = readTestFile(`public-${algorithm}-ssh`);
		if (
			keyFormat === KeyFormat.Pkcs1 ||
			keyFormat == KeyFormat.Sec1 ||
			keyFormat === KeyFormat.Pkcs8
		) {
			// Some formats don't support comments.
			const publicKey = await importKey(expected);
			publicKey.comment = null;
			expected = await exportPublicKey(publicKey);
		}

		const exported = await exportPublicKey(keyPair);
		assert.equal(exported, expected);
	}

	@test
	@params({ algorithm: 'rsa2048', keyFormat: KeyFormat.Pkcs1 })
	@params({ algorithm: 'rsa2048', keyFormat: KeyFormat.Pkcs8 })
	@params({ algorithm: 'rsa2048', keyFormat: KeyFormat.Pkcs8, passphrase: 'password' })
	@params({ algorithm: 'rsa2048', keyFormat: KeyFormat.Jwk })
	@params({ algorithm: 'rsa4096', keyFormat: KeyFormat.Pkcs1 })
	@params({ algorithm: 'rsa4096', keyFormat: KeyFormat.Pkcs8 })
	@params({ algorithm: 'rsa4096', keyFormat: KeyFormat.Jwk })
	@params({ algorithm: 'ecdsa384', keyFormat: KeyFormat.Sec1 })
	@params({ algorithm: 'ecdsa384', keyFormat: KeyFormat.Pkcs8 })
	@params({ algorithm: 'ecdsa384', keyFormat: KeyFormat.Pkcs8, passphrase: 'password' })
	@params({ algorithm: 'ecdsa384', keyFormat: KeyFormat.Jwk })
	@params({ algorithm: 'ecdsa521', keyFormat: KeyFormat.Sec1 })
	@params({ algorithm: 'ecdsa521', keyFormat: KeyFormat.Pkcs8 })
	@params.naming(
		(p) =>
			`importPrivateKey(${p.algorithm}, ${getFormatName(p.keyFormat)}, ${p.passphrase ??
				'null'})`,
	)
	public async importPrivateKey({
		algorithm,
		keyFormat,
		passphrase,
	}: {
		algorithm: string;
		keyFormat: KeyFormat;
		passphrase?: string;
	}) {
		const suffix = getFormatName(keyFormat) + (passphrase ? '-pw' : '');
		const keyPair = await importKey(readTestFile(`private-${algorithm}-${suffix}`), passphrase);
		assert(keyPair);
		assert(keyPair.hasPublicKey);
		assert(keyPair.hasPrivateKey);

		let expected = readTestFile(`public-${algorithm}-ssh`);
		if (
			keyFormat === KeyFormat.Pkcs1 ||
			keyFormat == KeyFormat.Sec1 ||
			keyFormat === KeyFormat.Pkcs8
		) {
			// Some formats don't support comments.
			const publicKey = await importKey(expected);
			publicKey.comment = null;
			expected = await exportPublicKey(publicKey);
		}

		try {
			const exported = await exportPublicKey(keyPair);
			assert.equal(exported, expected);
		} catch (e) {
			if (keyFormat === KeyFormat.Pkcs1) throw e;
		}
	}

	@test
	@pending(!process.platform) // Not implemented on browser platform.
	@params({ algorithm: 'rsa2048', keyFormat: KeyFormat.Pkcs1, passphrase: 'password' })
	@params({ algorithm: 'ecdsa384', keyFormat: KeyFormat.Sec1, passphrase: 'password' })
	@params.naming(
		(p) =>
			`importPrivateKey(${p.algorithm}, ${getFormatName(p.keyFormat)}, ${p.passphrase ??
				'null'})`,
	)
	public async importPrivateKey2({
		algorithm,
		keyFormat,
		passphrase,
	}: {
		algorithm: string;
		keyFormat: KeyFormat;
		passphrase?: string;
	}) {
		await this.importPrivateKey({ algorithm, keyFormat, passphrase });
	}

	@test
	@params({ algorithm: 'rsa2048', keyFormat: KeyFormat.Pkcs8 })
	@params({ algorithm: 'ecdsa384', keyFormat: KeyFormat.Pkcs8 })
	@params.naming(
		(p) => `importPrivateKeyInvalidPassword(${p.algorithm}, ${getFormatName(p.keyFormat)})`,
	)
	public async importPrivateKeyInvalidPassword({
		algorithm,
		keyFormat,
	}: {
		algorithm: string;
		keyFormat: KeyFormat;
	}) {
		const suffix = getFormatName(keyFormat) + '-pw';
		const privateKeyFile = readTestFile(`private-${algorithm}-${suffix}`);

		let error: Error | null = null;
		try {
			await importKey(privateKeyFile, null);
		} catch (e) {
			error = e as Error;
		}

		assert(error);
		assert.equal(error?.message, 'A passphrase is required to decrypt the key.');

		error = null;
		try {
			await importKey(privateKeyFile, 'invalid');
		} catch (e) {
			error = e as Error;
		}

		assert(error);
		assert.equal(error?.message, 'Key decryption failed - incorrect passphrase.');
	}

	@test
	@pending(!process.platform) // Not implemented on browser platform.
	@params({ algorithm: 'rsa2048', keyFormat: KeyFormat.Pkcs1 })
	@params({ algorithm: 'ecdsa384', keyFormat: KeyFormat.Sec1 })
	@params.naming(
		(p) => `importPrivateKeyInvalidPassword(${p.algorithm}, ${getFormatName(p.keyFormat)})`,
	)
	public async importPrivateKeyInvalidPassword2({
		algorithm,
		keyFormat,
	}: {
		algorithm: string;
		keyFormat: KeyFormat;
	}) {
		await this.importPrivateKeyInvalidPassword({ algorithm, keyFormat });
	}

	@test
	@params({ algorithm: 'rsa2048', keyFormat: KeyFormat.Ssh })
	@params({ algorithm: 'rsa2048', keyFormat: KeyFormat.Pkcs1 })
	@params({ algorithm: 'rsa2048', keyFormat: KeyFormat.Pkcs8 })
	@params({ algorithm: 'rsa2048', keyFormat: KeyFormat.Jwk })
	@params({ algorithm: 'rsa4096', keyFormat: KeyFormat.Ssh })
	@params({ algorithm: 'rsa4096', keyFormat: KeyFormat.Pkcs1 })
	@params({ algorithm: 'rsa4096', keyFormat: KeyFormat.Pkcs8 })
	@params({ algorithm: 'rsa4096', keyFormat: KeyFormat.Jwk })
	@params({ algorithm: 'ecdsa384', keyFormat: KeyFormat.Ssh })
	@params({ algorithm: 'ecdsa384', keyFormat: KeyFormat.Pkcs8 })
	@params({ algorithm: 'ecdsa384', keyFormat: KeyFormat.Jwk })
	@params({ algorithm: 'ecdsa521', keyFormat: KeyFormat.Ssh })
	@params({ algorithm: 'ecdsa521', keyFormat: KeyFormat.Pkcs8 })
	@params.naming((p) => `exportPublicKey(${p.algorithm}, ${getFormatName(p.keyFormat)})`)
	public async exportPublicKey({
		algorithm,
		keyFormat,
	}: {
		algorithm: string;
		keyFormat: KeyFormat;
	}) {
		const suffix = getFormatName(keyFormat);
		const publicKey = await importKey(readTestFile(`public-${algorithm}-ssh`));

		const exported = await exportPublicKey(publicKey, keyFormat);
		const expected = readTestFile(`public-${algorithm}-${suffix}`);
		assert.equal(exported, expected);
	}

	@test
	@params({ algorithm: 'ecdsa384', keyFormat: KeyFormat.Sec1 })
	@params.naming(
		(p) => `exportPublicKeyNotSupported(${p.algorithm}, ${getFormatName(p.keyFormat)})`,
	)
	public async exportPublicKeyNotSupported({
		algorithm,
		keyFormat,
	}: {
		algorithm: string;
		keyFormat: KeyFormat;
	}) {
		const suffix = getFormatName(keyFormat);
		const publicKey = await importKey(readTestFile(`public-${algorithm}-ssh`));

		let error: Error | null = null;
		try {
			await exportPublicKey(publicKey, keyFormat);
		} catch (e) {
			error = e as Error;
		}

		assert(error);
		assert(error?.message && /not supported/.test(error.message));
	}

	@test
	@params({ algorithm: 'rsa2048', keyFormat: KeyFormat.Pkcs1 })
	@params({ algorithm: 'rsa2048', keyFormat: KeyFormat.Pkcs1, passphrase: 'password' })
	@params({ algorithm: 'rsa2048', keyFormat: KeyFormat.Pkcs8 })
	@params({ algorithm: 'rsa2048', keyFormat: KeyFormat.Pkcs8, passphrase: 'password' })
	@params({ algorithm: 'rsa2048', keyFormat: KeyFormat.Jwk })
	@params({ algorithm: 'rsa4096', keyFormat: KeyFormat.Pkcs1 })
	@params({ algorithm: 'rsa4096', keyFormat: KeyFormat.Pkcs8 })
	@params({ algorithm: 'rsa4096', keyFormat: KeyFormat.Jwk })
	@params({ algorithm: 'ecdsa384', keyFormat: KeyFormat.Sec1 })
	@params({ algorithm: 'ecdsa384', keyFormat: KeyFormat.Sec1, passphrase: 'password' })
	@params({ algorithm: 'ecdsa384', keyFormat: KeyFormat.Pkcs8 })
	@params({ algorithm: 'ecdsa384', keyFormat: KeyFormat.Pkcs8, passphrase: 'password' })
	@params({ algorithm: 'ecdsa384', keyFormat: KeyFormat.Jwk })
	@params({ algorithm: 'ecdsa521', keyFormat: KeyFormat.Sec1 })
	@params({ algorithm: 'ecdsa521', keyFormat: KeyFormat.Pkcs8 })
	@params.naming(
		(p) =>
			`exportPrivateKey(${p.algorithm}, ${getFormatName(p.keyFormat)}, ${p.passphrase ??
				'null'})`,
	)
	public async exportPrivateKey({
		algorithm,
		keyFormat,
		passphrase,
	}: {
		algorithm: string;
		keyFormat: KeyFormat;
		passphrase?: string;
	}) {
		provideMockRandomBytes(keyFormat, !!passphrase);

		const suffix = getFormatName(keyFormat) + (passphrase ? '-pw' : '');
		const formatSuffix = algorithm.startsWith('rsa') ? 'pkcs1' : 'sec1';
		const keyPair = await importKey(readTestFile(`private-${algorithm}-${formatSuffix}`));
		keyPair.comment = 'comment';

		let error: Error | null = null;
		let exported: string | null = null;
		try {
			exported = await exportPrivateKey(keyPair, passphrase, keyFormat);
		} catch (e) {
			error = e as Error;
		}

		if (passphrase && (keyFormat === KeyFormat.Pkcs1 || keyFormat === KeyFormat.Sec1)) {
			// PKCS1/SEC1 export with passphrase is not supported due to weak encryption.
			assert(error);
			assert(error?.message && /export with passphrase is not supported/.test(error.message));
		} else {
			const expected = readTestFile(`private-${algorithm}-${suffix}`);
			assert.equal(exported, expected);
		}
	}

	@test
	@params({ algorithm: 'rsa2048', keyFormat: KeyFormat.Ssh })
	@params({ algorithm: 'rsa2048', keyFormat: KeyFormat.Pkcs1 })
	@params({ algorithm: 'rsa2048', keyFormat: KeyFormat.Pkcs8 })
	@params({ algorithm: 'rsa2048', keyFormat: KeyFormat.Jwk })
	@params({ algorithm: 'ecdsa384', keyFormat: KeyFormat.Ssh })
	@params({ algorithm: 'ecdsa384', keyFormat: KeyFormat.Pkcs8 })
	@params({ algorithm: 'ecdsa384', keyFormat: KeyFormat.Jwk })
	@params.naming(
		(p) => `exportImportPublicKeyBytes(${p.algorithm}, ${getFormatName(p.keyFormat)})`,
	)
	public async exportImportPublicKeyBytes({
		algorithm,
		keyFormat,
	}: {
		algorithm: string;
		keyFormat: KeyFormat;
	}) {
		const publicKey = await importKey(readTestFile(`public-${algorithm}-ssh`));
		const publicKeyBytes = await exportPublicKeyBytes(publicKey, keyFormat, KeyEncoding.Binary);
		const publicKey2 = await importKeyBytes(publicKeyBytes, null, keyFormat);
		publicKey2.comment = 'comment';

		const exported = await exportPublicKey(publicKey2, KeyFormat.Ssh);
		const expected = readTestFile(`public-${algorithm}-ssh`);
		assert.equal(exported, expected);
	}

	@test
	@params({ algorithm: 'rsa2048', keyFormat: KeyFormat.Pkcs1 })
	@params({ algorithm: 'rsa2048', keyFormat: KeyFormat.Pkcs8 })
	@params({ algorithm: 'rsa2048', keyFormat: KeyFormat.Pkcs8, passphrase: 'password' })
	@params({ algorithm: 'rsa2048', keyFormat: KeyFormat.Jwk })
	@params({ algorithm: 'ecdsa384', keyFormat: KeyFormat.Sec1 })
	@params({ algorithm: 'ecdsa384', keyFormat: KeyFormat.Pkcs8 })
	@params({ algorithm: 'ecdsa384', keyFormat: KeyFormat.Pkcs8, passphrase: 'password' })
	@params({ algorithm: 'ecdsa384', keyFormat: KeyFormat.Jwk })
	@params.naming(
		(p) =>
			`exportImportPrivateKeyBytes(${p.algorithm}, ${getFormatName(
				p.keyFormat,
			)}, ${p.passphrase ?? 'null'})`,
	)
	public async exportImportPrivateKeyBytes({
		algorithm,
		keyFormat,
		passphrase,
	}: {
		algorithm: string;
		keyFormat: KeyFormat;
		passphrase?: string;
	}) {
		provideMockRandomBytes(keyFormat, !!passphrase);

		const formatSuffix = algorithm.startsWith('rsa') ? 'pkcs1' : 'sec1';
		const privateKey = await importKey(readTestFile(`private-${algorithm}-${formatSuffix}`));
		const privateKeyBytes = await exportPrivateKeyBytes(
			privateKey,
			passphrase,
			keyFormat,
			KeyEncoding.Binary,
		);

		const privateKey2 = await importKeyBytes(privateKeyBytes, passphrase, keyFormat);

		const exportFormat = algorithm == 'rsa' ? KeyFormat.Pkcs1 : KeyFormat.Sec1;
		const exported = await exportPrivateKey(privateKey2, null, exportFormat);
		const expected = readTestFile(`private-${algorithm}-${formatSuffix}`);
		assert.equal(exported, expected);
	}
}

function readTestFile(name: string): string {
	const testKeys = require('./testKeys');
	const bytes: Buffer = testKeys[name];
	if (!bytes) throw new Error('Test key file not found: ' + name);
	return bytes.toString('utf8').replace(/\r/g, '');
}

function getFormatName(keyFormat: KeyFormat): string {
	switch (keyFormat) {
		case KeyFormat.Ssh:
			return 'ssh';
		case KeyFormat.Pkcs1:
			return 'pkcs1';
		case KeyFormat.Sec1:
			return 'sec1';
		case KeyFormat.Pkcs8:
			return 'pkcs8';
		case KeyFormat.Jwk:
			return 'jwk';
		default:
			throw new Error('Not implemented.');
	}
}

function provideMockRandomBytes(keyFormat: KeyFormat, encrypting: boolean): void {
	if (keyFormat === KeyFormat.Pkcs8 && encrypting) {
		var mockRandom = new MockRandom();
		var formatter = <Pkcs8KeyFormatter>keyFormatters.get(KeyFormat.Pkcs8);
		formatter.random = mockRandom;

		var salt = [0x1f, 0xc0, 0xf9, 0x60, 0xc9, 0x51, 0x89, 0x9a];
		mockRandom.values.push(Buffer.from(salt));

		// prettier-ignore
		var iv = [
			0x38, 0xaf, 0x3d, 0x00, 0xec, 0xe6, 0x43, 0x42,
			0x7c, 0x94, 0x30, 0x73, 0xcb, 0x92, 0x4f, 0xa2,
		];
		mockRandom.values.push(Buffer.from(iv));
	}
}

File: test/cs/Ssh.Test/KeyImportExportTests.cs
using System;
using System.IO;
using Microsoft.DevTunnels.Ssh.Keys;
using Xunit;

namespace Microsoft.DevTunnels.Ssh.Test;

public class KeyImportExportTests
{
	public KeyImportExportTests()
	{
		// Reduce the number of KDF rounds when testing so the tests run faster.
		var opensshFormatter = (OpenSshKeyFormatter)KeyPair.Formatters[KeyFormat.OpenSsh];
		opensshFormatter.KdfRounds = 2;

		var pkcs8Formatter = (Pkcs8KeyFormatter)KeyPair.Formatters[KeyFormat.Pkcs8];
		pkcs8Formatter.Pbkdf2Iterations = 2048;
	}

	private const string TestDataDir = "Data";
	private const string TestPassword = "password";

	private static string TestFile(string name)
	{
		var testFile = Path.Combine(TestDataDir, name + ".txt");
		if (!File.Exists(testFile))
		{
			throw new NotImplementedException($"Test file does not exist: {testFile}");
		}
		return testFile;
	}

	[SkippableTheory(typeof(NotImplementedException))]
	[InlineData("rsa2048", KeyFormat.Ssh)]
	[InlineData("rsa2048", KeyFormat.Ssh2)]
	[InlineData("rsa2048", KeyFormat.Pkcs1)]
	[InlineData("rsa2048", KeyFormat.Pkcs8)]
	[InlineData("rsa2048", KeyFormat.Jwk)]
	[InlineData("rsa4096", KeyFormat.Ssh)]
	[InlineData("rsa4096", KeyFormat.Ssh2)]
	[InlineData("rsa4096", KeyFormat.Pkcs1)]
	[InlineData("rsa4096", KeyFormat.Pkcs8)]
	[InlineData("ecdsa384", KeyFormat.Ssh)]
	[InlineData("ecdsa384", KeyFormat.Pkcs8)]
	[InlineData("ecdsa384", KeyFormat.Jwk)]
	[InlineData("ecdsa521", KeyFormat.Ssh)]
	public void ImportPublicKey(string algorithm, KeyFormat keyFormat)
	{
		var suffix = keyFormat.ToString().ToLowerInvariant();
		var keyPair = KeyPair.ImportKeyFile(TestFile($"testkey-public-{algorithm}-{suffix}"));
		Assert.NotNull(keyPair);
		Assert.False(keyPair.HasPrivateKey);

		var expected = File.ReadAllText(TestFile($"testkey-public-{algorithm}-ssh"));
		if (keyFormat == KeyFormat.Pkcs1 ||
			keyFormat == KeyFormat.Sec1 ||
			keyFormat == KeyFormat.Pkcs8)
		{
			// Some formats don't support comments.
			var publicKey = KeyPair.ImportKey(expected);
			publicKey.Comment = null;
			expected = KeyPair.ExportPublicKey(publicKey);
		}

		Assert.Equal(expected, KeyPair.ExportPublicKey(keyPair));
	}

	[SkippableTheory(typeof(NotImplementedException))]
	[InlineData("rsa2048", KeyFormat.Pkcs1)]
	[InlineData("rsa2048", KeyFormat.Pkcs8)]
	[InlineData("rsa2048", KeyFormat.Ssh2)]
	[InlineData("rsa2048", KeyFormat.OpenSsh)]
	[InlineData("rsa2048", KeyFormat.Jwk)]
	[InlineData("rsa2048", KeyFormat.Pkcs1, TestPassword)]
	[InlineData("rsa2048", KeyFormat.Ssh2, TestPassword)]
	[InlineData("rsa2048", KeyFormat.OpenSsh, TestPassword)]
	[InlineData("rsa4096", KeyFormat.Pkcs1)]
	[InlineData("rsa4096", KeyFormat.Pkcs8)]
	[InlineData("rsa4096", KeyFormat.Ssh2)]
	[InlineData("rsa4096", KeyFormat.OpenSsh)]
	[InlineData("rsa4096", KeyFormat.Jwk)]
	[InlineData("ecdsa384", KeyFormat.Sec1)]
	[InlineData("ecdsa384", KeyFormat.Sec1, TestPassword)]
	[InlineData("ecdsa384", KeyFormat.Pkcs8)]
	[InlineData("ecdsa384", KeyFormat.OpenSsh)]
	[InlineData("ecdsa384", KeyFormat.OpenSsh, TestPassword)]
	[InlineData("ecdsa384", KeyFormat.Jwk)]
	[InlineData("ecdsa521", KeyFormat.Sec1)]
	[InlineData("ecdsa521", KeyFormat.Pkcs8)]
#if SSH_ENABLE_PBKDF2
	[InlineData("rsa2048", KeyFormat.Pkcs8, TestPassword)]
	[InlineData("ecdsa384", KeyFormat.Pkcs8, TestPassword)]
#endif
	public void ImportPrivateKey(string algorithm, KeyFormat keyFormat, string passphrase = null)
	{
		var suffix = keyFormat.ToString().ToLowerInvariant()
			+ (passphrase != null ? "-pw" : string.Empty);

		var keyPair = KeyPair.ImportKeyFile(
			TestFile($"testkey-private-{algorithm}-{suffix}"), passphrase);
		Assert.NotNull(keyPair);
		Assert.True(keyPair.HasPrivateKey);

		var expected = File.ReadAllText(TestFile($"testkey-public-{algorithm}-ssh"));
		if (keyFormat == KeyFormat.Pkcs1 ||
			keyFormat == KeyFormat.Sec1 ||
			keyFormat == KeyFormat.Pkcs8)
		{
			// Some formats don't support comments.
			var publicKey = KeyPair.ImportKey(expected);
			publicKey.Comment = null;
			expected = KeyPair.ExportPublicKey(publicKey);
		}

		Assert.Equal(expected, KeyPair.ExportPublicKey(keyPair));
	}

	[SkippableTheory(typeof(NotImplementedException))]
	[InlineData("rsa2048", KeyFormat.Pkcs1)]
	[InlineData("rsa2048", KeyFormat.Ssh2)]
	[InlineData("rsa2048", KeyFormat.OpenSsh)]
	[InlineData("ecdsa384", KeyFormat.OpenSsh)]
#if SSH_ENABLE_PBKDF2
	[InlineData("rsa2048", KeyFormat.Pkcs8)]
	[InlineData("ecdsa384", KeyFormat.Pkcs8)]
#endif
	public void ImportPrivateKeyInvalidPassword(string algorithm, KeyFormat keyFormat)
	{
		var suffix = keyFormat.ToString().ToLowerInvariant() + "-pw";
		var privateKeyFile = TestFile($"testkey-private-{algorithm}-{suffix}");

		Assert.Throws<UnauthorizedAccessException>(() =>
		{
			KeyPair.ImportKeyFile(privateKeyFile, null);
		});
		Assert.Throws<UnauthorizedAccessException>(() =>
		{
			KeyPair.ImportKeyFile(privateKeyFile, "invalid");
		});
	}

	[SkippableTheory(typeof(NotImplementedException))]
	[InlineData("rsa2048", KeyFormat.Ssh)]
	[InlineData("rsa2048", KeyFormat.Ssh2)]
	[InlineData("rsa2048", KeyFormat.Pkcs1)]
	[InlineData("rsa2048", KeyFormat.Pkcs8)]
	[InlineData("rsa2048", KeyFormat.Jwk)]
	[InlineData("rsa4096", KeyFormat.Ssh)]
	[InlineData("rsa4096", KeyFormat.Ssh2)]
	[InlineData("rsa4096", KeyFormat.Pkcs1)]
	[InlineData("rsa4096", KeyFormat.Pkcs8)]
	[InlineData("rsa4096", KeyFormat.Jwk)]
	[InlineData("ecdsa384", KeyFormat.Ssh)]
	[InlineData("ecdsa384", KeyFormat.Pkcs8)]
	[InlineData("ecdsa384", KeyFormat.Jwk)]
	[InlineData("ecdsa521", KeyFormat.Ssh)]
	[InlineData("ecdsa521", KeyFormat.Pkcs8)]
	public void ExportPublicKey(string algorithm, KeyFormat keyFormat)
	{
		var suffix = keyFormat.ToString().ToLowerInvariant();
		var publicKey = KeyPair.ImportKeyFile(TestFile($"testkey-public-{algorithm}-ssh"));

		var exportedPublicKey = KeyPair.ExportPublicKey(publicKey, keyFormat);
		var expected = File.ReadAllText(TestFile($"testkey-public-{algorithm}-{suffix}"));
		Assert.Equal(expected, exportedPublicKey);
	}

	[SkippableTheory(typeof(NotImplementedException))]
	[InlineData("ecdsa384", KeyFormat.Sec1)]
	public void ExportPublicKeyNotSupported(string algorithm, KeyFormat keyFormat)
	{
		var suffix = keyFormat.ToString().ToLowerInvariant();
		var publicKey = KeyPair.ImportKeyFile(TestFile($"testkey-public-{algorithm}-ssh"));
		Assert.Throws<NotSupportedException>(() => KeyPair.ExportPublicKey(publicKey, keyFormat));
	}

	[SkippableTheory(typeof(NotImplementedException))]
	[InlineData("rsa2048", KeyFormat.Pkcs1)]
	[InlineData("rsa2048", KeyFormat.Pkcs8)]
	[InlineData("rsa2048", KeyFormat.Ssh2)]
	[InlineData("rsa2048", KeyFormat.OpenSsh)]
	[InlineData("rsa2048", KeyFormat.Jwk)]
	[InlineData("rsa2048", KeyFormat.Pkcs1, TestPassword)]
	[InlineData("rsa2048", KeyFormat.Ssh2, TestPassword)]
	[InlineData("rsa2048", KeyFormat.OpenSsh, TestPassword)]
	[InlineData("rsa4096", KeyFormat.Pkcs1)]
	[InlineData("rsa4096", KeyFormat.Pkcs8)]
	[InlineData("rsa4096", KeyFormat.Ssh2)]
	[InlineData("rsa4096", KeyFormat.OpenSsh)]
	[InlineData("rsa4096", KeyFormat.Jwk)]
	[InlineData("ecdsa384", KeyFormat.Sec1)]
	[InlineData("ecdsa384", KeyFormat.Sec1, TestPassword)]
	[InlineData("ecdsa384", KeyFormat.Pkcs8)]
	[InlineData("ecdsa384", KeyFormat.OpenSsh)]
	[InlineData("ecdsa384", KeyFormat.OpenSsh, TestPassword)]
	[InlineData("ecdsa521", KeyFormat.Sec1)]
	[InlineData("ecdsa521", KeyFormat.Pkcs8)]
#if SSH_ENABLE_PBKDF2
	[InlineData("rsa2048", KeyFormat.Pkcs8, TestPassword)]
	[InlineData("ecdsa384", KeyFormat.Pkcs8, TestPassword)]
#endif
	public void ExportPrivateKey(string algorithm, KeyFormat keyFormat, string passphrase = null)
	{
		ProvideMockRandomBytes(keyFormat, passphrase != null);

		var suffix = keyFormat.ToString().ToLowerInvariant()
			+ (passphrase != null ? "-pw" : string.Empty);
		var formatSuffix = algorithm.StartsWith("rsa") ? "pkcs1" : "sec1";

		var keyPair = KeyPair.ImportKeyFile(
			TestFile($"testkey-private-{algorithm}-{formatSuffix}"));
		keyPair.Comment = "comment"; // PKCS#1/SEC1 format does not persist the comment.

		if (passphrase != null && keyFormat != KeyFormat.Pkcs8 && keyFormat != KeyFormat.OpenSsh)
		{
			// Export with passphrase is not supported due to weak encryption.
			Assert.Throws<NotSupportedException>(
				() => KeyPair.ExportPrivateKey(keyPair, passphrase, keyFormat));
		}
		else
		{
			var exportedPrivateKey = KeyPair.ExportPrivateKey(keyPair, passphrase, keyFormat);
			var expected = File.ReadAllText(TestFile($"testkey-private-{algorithm}-{suffix}"));
			Assert.Equal(expected, exportedPrivateKey);
		}
	}

	[SkippableTheory(typeof(NotImplementedException))]
	[InlineData("rsa2048", KeyFormat.Ssh)]
	[InlineData("rsa2048", KeyFormat.Ssh2)]
	[InlineData("rsa2048", KeyFormat.Pkcs1)]
	[InlineData("rsa2048", KeyFormat.Pkcs8)]
	[InlineData("ecdsa384", KeyFormat.Ssh)]
	[InlineData("ecdsa384", KeyFormat.Pkcs8)]
	public void ExportImportPublicKeyBytes(string algorithm, KeyFormat keyFormat)
	{
		var publicKey = KeyPair.ImportKeyFile(TestFile($"testkey-public-{algorithm}-ssh"));
		var publicKeyBytes = KeyPair.ExportPublicKeyBytes(
			publicKey, keyFormat, KeyEncoding.Binary);
		var publicKey2 = KeyPair.ImportKeyBytes(publicKeyBytes, passphrase: null, keyFormat);
		publicKey2.Comment = "comment";

		var exportedPublicKey = KeyPair.ExportPublicKey(publicKey2, KeyFormat.Ssh);
		var expected = File.ReadAllText(TestFile($"testkey-public-{algorithm}-ssh"));
		Assert.Equal(expected, exportedPublicKey);
	}

	[SkippableTheory(typeof(NotImplementedException))]
	[InlineData("rsa2048", KeyFormat.Pkcs1)]
	[InlineData("rsa2048", KeyFormat.Pkcs8)]
	[InlineData("rsa2048", KeyFormat.Ssh2)]
	[InlineData("rsa2048", KeyFormat.OpenSsh)]
	[InlineData("rsa2048", KeyFormat.OpenSsh, TestPassword)]
	[InlineData("ecdsa384", KeyFormat.Sec1)]
	[InlineData("ecdsa384", KeyFormat.Pkcs8)]
	[InlineData("ecdsa384", KeyFormat.OpenSsh)]
	[InlineData("ecdsa384", KeyFormat.OpenSsh, TestPassword)]
#if SSH_ENABLE_PBKDF2
	[InlineData("rsa2048", KeyFormat.Pkcs8, TestPassword)]
	[InlineData("ecdsa384", KeyFormat.Pkcs8, TestPassword)]
#endif
	public void ExportImportPrivateKeyBytes(
		string algorithm,
		KeyFormat keyFormat,
		string passphrase = null)
	{
		var formatSuffix = algorithm.StartsWith("rsa") ? "pkcs1" : "sec1";
		var privateKey = KeyPair.ImportKeyFile(TestFile(
			$"testkey-private-{algorithm}-{formatSuffix}"));

		ProvideMockRandomBytes(keyFormat, passphrase != null);
		var privateKeyBytes = KeyPair.ExportPrivateKeyBytes(
			privateKey, passphrase, keyFormat, KeyEncoding.Binary);

		var privateKey2 = KeyPair.ImportKeyBytes(privateKeyBytes, passphrase, keyFormat);

		var exportFormat = algorithm.StartsWith("rsa") ? KeyFormat.Pkcs1 : KeyFormat.Sec1;
		var exportedPrivateKey = KeyPair.ExportPrivateKey(
			privateKey2, passphrase: null, exportFormat);
		var expected = File.ReadAllText(TestFile($"testkey-private-{algorithm}-{formatSuffix}"));
		Assert.Equal(expected, exportedPrivateKey);
	}

	/// <summary>
	/// Provide mocked "random" bytes to exporters to ensure deterministic output.
	/// </summary>
	private static void ProvideMockRandomBytes(KeyFormat keyFormat, bool encrypting)
	{
		if (keyFormat == KeyFormat.OpenSsh)
		{
			var mockRandom = new MockRandom();
			var formatter = (OpenSshKeyFormatter)KeyPair.Formatters[KeyFormat.OpenSsh];
			formatter.Random = mockRandom;

			var checkBytes = new byte[] { 0xcc, 0xcc, 0xcc, 0xcc };
			mockRandom.Values.Add(checkBytes);

			if (encrypting)
			{
				var salt = new byte[]
				{
						0xc4, 0xdf, 0xeb, 0xf5, 0xd0, 0xb8, 0x03, 0x28,
						0xee, 0xf0, 0x07, 0xcd, 0xf1, 0xd6, 0x4b, 0xa1,
				};
				mockRandom.Values.Add(salt);
			}
		}
		else if (keyFormat == KeyFormat.Pkcs8 && encrypting)
		{
			var mockRandom = new MockRandom();
			var formatter = (Pkcs8KeyFormatter)KeyPair.Formatters[KeyFormat.Pkcs8];
			formatter.Random = mockRandom;

			var salt = new byte[]
			{
					0x1f, 0xc0, 0xf9, 0x60, 0xc9, 0x51, 0x89, 0x9a,
			};
			mockRandom.Values.Add(salt);

			var iv = new byte[]
			{
					0x38, 0xaf, 0x3d, 0x00, 0xec, 0xe6, 0x43, 0x42,
					0x7c, 0x94, 0x30, 0x73, 0xcb, 0x92, 0x4f, 0xa2,
			};
			mockRandom.Values.Add(iv);
		}
	}
}
